util/gc/gcfst.py

```diff
@@ -0,0 +1,197 @@
+import struct
+
+from segtypes.gc.segment import GCSegment
+from pathlib import Path
+from util import options
+from typing import List, Optional
+
+
+# Represents the info for either a directory or a file within a GameCube disc image's file system.
+class GCFSTEntry:
+    def __init__(self, flags: bool, name_offset, offset, length):
+        self.flags = flags
+        self.name_offset = name_offset
+        self.offset = offset
+        self.length = length
+
+        self.name = ""
+        self.parent: Optional[GCFSTEntry] = None
+        self.children: List[GCFSTEntry] = []
+
+    def populate_children_recursive(
+        self, root_dir: "GCFSTEntry", current_node_offset, fst_bytes, string_table_bytes
+    ):
+        self.parent = root_dir
+        self.read_name(string_table_bytes)
+
+        # This node is a file, so we don't do anything but return that we read 1 node.
+        if self.flags == False:
+            return 1
+
+        nodes_read = 1
+        next_child_offset = current_node_offset + 0x0C
+
+        # Directory nodes contain the index of the next node that is NOT its child, meaning the index of their next sibling node.
+        # We can figure out when we're done reading child nodes by comparing the offset of the next node to read to the
+        # offset of the next sibling node. We stop reading when the next node offset is >= the offset of the next sibling node.
+        while next_child_offset < self.length * 0x0C:
+            new_entry = GCFSTEntry(
+                bool(fst_bytes[next_child_offset + 0x0000]),
+                struct.unpack(
+                    ">I", fst_bytes[next_child_offset : next_child_offset + 0x0004]
+                )[0]
+                & 0x00FFFFFF,
+                struct.unpack(
+                    ">I",
+                    fst_bytes[next_child_offset + 0x0004 : next_child_offset + 0x0008],
+                )[0],
+                struct.unpack(
+                    ">I",
+                    fst_bytes[next_child_offset + 0x0008 : next_child_offset + 0x000C],
+                )[0],
+            )
+
+            self.children.append(new_entry)
+            nodes_read += new_entry.populate_children_recursive(
+                self, next_child_offset, fst_bytes, string_table_bytes
+            )
+
+            next_child_offset = current_node_offset + nodes_read * 0x0C
+
+        return nodes_read
+
+    # Reads the name of this FST entry from the given bytes array.
+    def read_name(self, string_table_bytes):
+        offset = 0
+        bytes = []
+
+        for offset in range(len(string_table_bytes) - self.name_offset):
+            cur_byte = string_table_bytes[self.name_offset + offset]
+            if cur_byte == 0x00:
+                break
+
+            bytes.append(cur_byte)
+
+        self.name = bytearray(bytes).decode("shift-jis")
```

Comment:
Isn't this the same function as the one in `rarc.py`? In that case it would be better to move this to the util folder instead of duplicating the code